src/tribol/mesh/MfemData.hpp

```diff
@@ -0,0 +1,1390 @@
+
+  /**
+   * @brief Grid function on the submesh
+   *
+   * @note Stored as a reference wrapper so the reference can be updated
+   */
+  std::reference_wrapper<const mfem::ParGridFunction> submesh_;
+
+  /**
+   * @brief UpdateData object created upon call to UpdateField()
+   */
+  std::unique_ptr<UpdateData> update_data_;
+};
+
+/**
+ * @brief Stores MFEM and transfer data associated with parent vector fields
+ * (displacement and velocity)
+ */
+class MfemMeshData
+{
+public:
+  /**
+   * @brief Construct a new MfemMeshData object
+   * 
+   * @param mesh_id_1 Integer identifier for first Tribol registered mesh
+   * @param mesh_id_2 Integer identifier for second Tribol registered mesh
+   * @param mesh Volume mesh of parent domain
+   * @param current_coords Grid function (on mesh) holding current coordinates
+   * @param attributes_1 Mesh boundary attributes identifying first mesh
+   * @param attributes_2 Mesh boundary attributes identifying second mesh
+   */
+  MfemMeshData(
+    integer mesh_id_1,
+    integer mesh_id_2,
+    mfem::ParMesh& mesh,
+    const mfem::ParGridFunction& current_coords,
+    const std::set<integer>& attributes_1,
+    const std::set<integer>& attributes_2
+  );
+
+  /**
+   * @brief Get coordinate grid function on the parent mesh
+   * 
+   * @return const mfem::ParGridFunction& 
+   */
+  const mfem::ParGridFunction& GetParentCoords() const
+  {
+    return coords_.GetParentField();
+  }
+
+  /**
+   * @brief Sets a new coordinate grid function on the parent mesh
+   * 
+   * @param current_coords Coordinate grid function on the parent mesh
+   */
+  void SetParentCoords(const mfem::ParGridFunction& current_coords);
+
+  /**
+   * @brief Build a new redecomp mesh and update grid functions
+   *
+   * @note This method should be called after the coordinate grid function has
+   * changed.
+   */
+  void UpdateMeshData();
+
+  /**
+   * @brief Get the integer identifier for the first Tribol registered mesh
+   *
+   * @return integer 
+   */
+  integer GetMesh1ID() const { return mesh_id_1_; }
+
+  /**
+   * @brief Get the integer identifier for the second Tribol registered mesh
+   *
+   * @return integer 
+   */
+  integer GetMesh2ID() const { return mesh_id_2_; }
+
+  /**
+   * @brief Get the number of elements on the first mesh
+   * 
+   * @return integer 
+   */
+  integer GetMesh1NE() const
+  { 
+    return GetUpdateData().conn_1_.size() / num_verts_per_elem_;
+  }
+
+  /**
+   * @brief Get the number of elements on the second mesh
+   * 
+   * @return integer 
+   */
+  integer GetMesh2NE() const
+  {
+    return GetUpdateData().conn_2_.size() / num_verts_per_elem_;
+  }
+
+  /**
+   * @brief Get the total number of vertices in both meshes
+   * 
+   * @return integer 
+   */
+  integer GetNV() const { return GetUpdateData().redecomp_.GetNV(); }
+
+  /**
+   * @brief Get the connectivity for the first mesh
+   * 
+   * @return const IndexType* 
+   */
+  const IndexType* GetMesh1Conn() const
```

Comment:
first mesh = first surface submesh?

Reply:
Based on the hierarchy we discussed (`parent mesh`, `submesh`, `LOR mesh`, `redecomp mesh`), a shorthand for the Tribol registered mesh in variable names could be `mesh`?  Thoughts @srwopschall?  In which case, we wouldn't need to change this.